rfcs/20231017-collective-broadcast.md

```diff
@@ -0,0 +1,109 @@
+
+Additionally, a new compiler pass that detects usage of the old `psum` hack and replaces it with a `collective_broadcast` could be implemented only once and forever be supported by all hardware, future and current. This could have positive knock-on effects for users who don't even realize they're using it!
+
+`collective_broadcast` can be used to quickly replicate a tensor across an entire mesh, and would use less communication resources as compared to `all_gather` or `psum`. `collective_broadcast` is also the base primitive used in the [SUMMA](https://www.netlib.org/lapack/lawnspdf/lawn96.pdf) distributed GEMM algorithm. As AI computing grows larger, there likely will grow a need for these 2D distributed GEMM algorithms. Adding support for one of the needed primitives could help advance research in these areas. 
+
+## Alternatives considered
+
+Instead of adding `collective_broadcast` as a primitive, we considered loosening the restriction of `collective_permute` to allow a one-to-many communication schedule instead of the current restriction of a one-to-one schedule. Downstream compilers would then be responsible for detecting this and calling their own `collective_broadcast` primitive. However, loosening this restriction makes defining the transposition rule for `collective_permute` significantly more complicated. Questions of how to calculate that and do it efficiently given any communication configuration and do so in SPMD became difficult. However, the transposition rule for `collective_broadcast` is just `psum` with a source-device one-hot masking. This simplicity plus the broad usage of `collective_broadcast` in the wider ecosystem made us choose to ultimately add the new primitive instead.
+
+## Why call it collective_broadcast and not just broadcast?
+Unfortunately, the op name `broadcast` is already taken by [an op in XLA proper](https://www.tensorflow.org/xla/operation_semantics#broadcast), so we can't have the two names clash. `collective_broadcast` was the preferred alternative.
+
+## Proposed Specification
+
+### collective_broadcast
+
+#### Semantics
+
+Within each process group in the StableHLO process grid, send the value of the
+`operand` tensor from the source process to the target processes and produce a
+`result` tensor. For any process `n` not in any replica group, that process will just
+return zeros.
+
+The operation splits the StableHLO process grid into `process_groups` which is
+defined as follows:
+
+* `cross_replica(replica_groups)` if `channel_id <= 0`.
```

Comment:
Sorry I'm new here, so I'm not sure this is a comment on the doc or an actual question :)

What's does having a channel_id <= 0 here mean? This sounds maybe like an attribute / argument?

Reply:
`channel_id <= 0 ` means that this attribute is not provided (See [StablehloAttrs.td](https://github.com/openxla/stablehlo/blob/654c202babbab3f4bd50cdd276de02dc17ecc358/stablehlo/dialect/StablehloAttrs.td#L113))

Reply:
What do you think about making it an error then? 

Also, can you update this doc to say channel_id <= 0 means this attribute isn't provided?

Reply:
> What do you think about making it an error then?
> Also, can you update this doc to say channel_id <= 0 means this attribute isn't provided?

I think it's the other way around. `channel_id` is not needed for cross replica communication (e.g. happening within an accelerator, and can communicate directly with other cores) whereas it is required for cross partition (e.g. happening among multiple accelerators and thus require a unique "channel" to communicate amongst each other). So `channel_id = 0` is not because this attribute is not provided, but rather because we are doing cross replica communication, so `channel_id` is not needed and it defaults to a value like 0.

However, whether negative ids are allowed, that's a broader question outside the scope of this RFC. IIRC `channel_id` being signed is a legacy inherited from MHLO.

Reply:
Ack, thanks for the thorough explanation :)